apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java

```diff
@@ -116,43 +120,27 @@ public ApikeyController(ApikeyRepo apikeyRepo) {
-                                   Tools.generatePassPhrase(10),
-                                   apikeyCreate.getFirstName(),
-                                   apikeyCreate.getLastName(),
-                                   apikeyCreate.getEmail(),
-                                   null != apikeyCreate.getLevel() && apikeyCreate.getLevel()
-                                            .equalsIgnoreCase(Level.ADMIN.getLevelName()) ?
-                                            Level.ADMIN.getLevelName() : Level.CLIENT.getLevelName());
-        if (null != apikeyCreate.getWebsite()) {
-            apikey.setWebsite(apikeyCreate.getWebsite());
-        }
-        if (null != apikeyCreate.getAppName()) {
-            apikey.setAppName(apikeyCreate.getAppName());
-        }
-        if (null != apikeyCreate.getCompany()) {
-            apikey.setCompany(apikeyCreate.getCompany());
-        }
-        if (null != apikeyCreate.getSector()) {
-            apikey.setSector(apikeyCreate.getSector());
+        KeycloakAuthenticationToken keycloakAuthenticationToken = (KeycloakAuthenticationToken) SecurityContextHolder.getContext().getAuthentication();
+        KeycloakSecurityContext securityContext = (KeycloakSecurityContext) keycloakAuthenticationToken.getCredentials();
+        try {
+            FullApikey apikey = keycloakManager.createClient(securityContext, apikeyCreate);
+            if (apikey != null) {
+                this.apikeyRepo.save(new Apikey(apikey));
+                LOG.debug("apikey: {} created", apikey.getApikey());
+
+                emailService.sendSimpleMessageUsingTemplate(apikey.getEmail(),
+                        "Your Europeana API keys",
+                        apikeyCreatedMail,
+                        apikey.getFirstName(),
+                        apikey.getLastName(),
+                        apikey.getApikey(),
+                        apikey.getClientSecret());
+                return new ResponseEntity<>(apikey, HttpStatus.CREATED);
+            }
+        } catch (ApikeyException e) {
+            return new ResponseEntity<>(e, HttpStatus.valueOf(e.getStatus()));
         }
-        this.apikeyRepo.save(apikey);
-        LOG.debug("apikey: {} created", apikey.getApikey());
-
-        emailService.sendSimpleMessageUsingTemplate(apikey.getEmail(),
-                                                    "Your Europeana API keys",
-                                                    apikeyCreatedMail,
-                                                    apikey.getFirstName(),
-                                                    apikey.getLastName(),
-                                                    apikey.getApikey(),
-                                                    apikey.getPrivatekey());
-        return new ResponseEntity<>(apikey, HttpStatus.CREATED);
+        return new ResponseEntity<>(new ApikeyException(400, MISSINGPARAMETER, missing), HttpStatus.BAD_REQUEST);
```

Comment:
Also ... the generation of the missing params list can be simplified by using list.toString();

Reply:
> Also ... the generation of the missing params list can be simplified by using list.toString();

It could be difficult I think - there are different words between each value from the list (",", "and")

Reply:
I added the comment here, as the implementation of the validateMissing() was not touch by this pool request, but the comment regarding missing params was related to the implementation of that method, which has a lot of redundant if,elses. Building the list of missing properties and serialization with list.toString() should be sufficient (the toString returns a json-like serialization for lists...)